### PR TITLE
Credit card validator 66

### PR DESCRIPTION
**Pull request by wr-candidate**

### CODE REVIEW

File: .classpath [hunk 1]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (wr-candidate) [inline .classpath:1]: Avoid committing IDE-specific files.

File: .project [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (wr-candidate) [inline .project:1]: ditto

File: .settings/org.eclipse.jdt.core.prefs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+eclipse.preferences.version=1
```
Comment (wr-candidate) [inline .settings/org.eclipse.jdt.core.prefs:1]: ditto

File: src/validator/CreditCardChecking.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package validator;
```
Comment (wr-candidate) [inline src/validator/CreditCardChecking.java:1]: Make package name more specific with organization prefix, eg `com.wr.cc.validator`

File: src/validator/CreditCardChecking.java [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+
+public interface CreditCardChecking {
```
Comment (wr-candidate) [inline src/validator/CreditCardChecking.java:3]: Classes should be named using nouns, eg `CreditCardChecker`. Also - `Validator` might be a more suitable name.

File: src/validator/CreditCardChecking.java [hunk 6]
```diff
@@ -0,0 +1,5 @@
...
+	public boolean checkValidity(CreditCardVendor ccv, String number);
```
Comment (wr-candidate) [inline src/validator/CreditCardChecking.java:4]: 1. `checkValidity` -> `validate` in short
2. add `final` to variables to indicate immutability
3. possibly (depending on the further implementation changes) a `isApplicable` or `getVendorName` method might be suitable

File: Base_of_credit_card_vendors.txt [hunk 7]
```diff
@@ -0,0 +1,2 @@
+Visa 13/16 4
```
Comment (wr-candidate) [inline Base_of_credit_card_vendors.txt:1]: Please use human-redable format to store configuration details. In this form the configuration is unclear without reading the code. `YAML` is widely used format you can consider.

File: src/validator/CreditCardValidatorProject.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            }catch(NumberFormatException | IOException e){
+                System.err.println("Invalid Format!");
+            }
+        	
+        }
+        
+    }
+	
+	static private void printMenu(){
+		System.out.println("Select your card vendor from list below("+vendors.size()+" exits):");
```
Comment (wr-candidate) [inline src/validator/CreditCardValidatorProject.java:48]: General comment: the code formatting in this PR is inconsistent. Please use IDE configured in a team-agreed way so that the formatting is consistent.

File: src/validator/Tests.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+import static org.junit.Assert.*;
+
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+
+import org.junit.Test;
+
+public class Tests {
```
Comment (wr-candidate) [inline src/validator/Tests.java:11]: The test classes should be named after the classes under test. In this case it should be `CreditCardValidatorTest`. That way it is easy to find test cases for the specific class.

File: src/validator/Tests.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+	private static CreditCardVendorsReader reader = new CreditCardVendorsReader();
+	private static ArrayList<CreditCardVendor> vendors = reader.returnVendorArray();
+	private static CreditCardValidator validator = new CreditCardValidator();
+	private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+	private static CreditCardVendor vendorMasterCard = vendors.get(1);
```
Comment (wr-candidate) [inline src/validator/Tests.java:16]: This relies on:
1. production configuration (once changed the tests might fail) - please use mocking instead
2. _magic numbers_ - it should be possible to reference vendor by a human-readable identifies

File: src/validator/Tests.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+
+import org.junit.Test;
+
+public class Tests {
+	private static CreditCardVendorsReader reader = new CreditCardVendorsReader();
+	private static ArrayList<CreditCardVendor> vendors = reader.returnVendorArray();
+	private static CreditCardValidator validator = new CreditCardValidator();
+	private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
```
Comment (wr-candidate) [inline src/validator/Tests.java:15]: unused

File: src/validator/Tests.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	private static CreditCardVendor vendorMasterCard = vendors.get(1);
+	
+
+	@Test
+	public void testLengthOfNumberWhenCorrect() {
```
Comment (wr-candidate) [inline src/validator/Tests.java:20]: All of those tests rely on the knowledge of the validator internals. Instead the tests should focus on checking if particular CC number is valid or not for given vendor. And then some edge cases should be checked (too short, too long, invalid prefix).

File: src/validator/Tests.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+	}
+	
+	@Test
+	public void testLuhnAlghoritmWhenCorrect() {
+		String number = "5584239583699571";
+		assertEquals(true, validator.checkValidity(vendorMasterCard, number));
+	}
+	
+	@Test
+	public void testLuhnAlghoritmWhenIncorrect() {
```
Comment (wr-candidate) [inline src/validator/Tests.java:50]: In order to validate a specific part of the CC validation algorithm, please extract those parts to a separate classes and write tests for those parts individually.

File: src/validator/Tests.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+	
+
+	@Test
+	public void testLengthOfNumberWhenCorrect() {
+		String number = "5584239583699571";
+		assertEquals(true, validator.checkValidity(vendorMasterCard, number));
+	}
+	
+	@Test
+	public void testLengthOfNumberWhenIncorrect() {
```
Comment (wr-candidate) [inline src/validator/Tests.java:26]: Consider using test input providers to avoid code duplication.

File: src/validator/CreditCardVendorsReader.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+
+public class CreditCardVendorsReader implements CreditCardVendorReading {
+	@Override
+	public ArrayList<CreditCardVendor> returnVendorArray() {
+		ArrayList<CreditCardVendor> list = new ArrayList<CreditCardVendor>();
+		
+		try(BufferedReader br = new BufferedReader(new FileReader("Base_of_credit_card_vendors.txt"))) {
```
Comment (wr-candidate) [inline src/validator/CreditCardVendorsReader.java:14]: _Magic string_ - the configuration file name should be at least moved to constant, or--ideally--provided as a constructor parameter.

File: src/validator/CreditCardVendorsReader.java [hunk 16]
```diff
@@ -0,0 +1,58 @@
+package validator;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+
+public class CreditCardVendorsReader implements CreditCardVendorReading {
```
Comment (wr-candidate) [inline src/validator/CreditCardVendorsReader.java:9]: This class has 3 responsibilities:
- configuration reading
- configuration parsing
- vendor instantiation

Please split this class into 3 separate classes (eg.: `ConfigReader`, `VendorFactory` and `VendorBuilder`).

File: src/validator/CreditCardVendorReading.java [hunk 17]
```diff
@@ -0,0 +1,7 @@
+package validator;
+
+import java.util.ArrayList;
+
+public interface CreditCardVendorReading {
```
Comment (wr-candidate) [inline src/validator/CreditCardVendorReading.java:5]: Unused once `CreditCardVendorsReader` will be refactored into 3 separate classes.

File: src/validator/CreditCardVendor.java [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+	}
+	
+	public CreditCardVendor(String vendorName, ArrayList<Integer> charsNumbers, ArrayList<String> masks){
+		name = vendorName;
+		numberOfChars = charsNumbers;
+		IIN = masks;
+	}
+	
+	@Override
+	public String toString(){
```
Comment (wr-candidate) [inline src/validator/CreditCardVendor.java:29]: Incomplete implementation. Also `hashCode` and `equals` is missing. What is more the class is mutable, while it should be immutable.
Please consider using [Immutables](https://immutables.github.io/) to introduce immutability. As a benefit correct `toString`, `equals` and `hashCode` will be generated as well.

File: src/validator/CreditCardVendor.java [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+	
+	public ArrayList<Integer> getNumberOfChars(){
+		return numberOfChars;
+	}
+	
+	public ArrayList<String> getIIN(){
+		return IIN;
+	}
+	
+	public CreditCardVendor(String vendorName, ArrayList<Integer> charsNumbers, ArrayList<String> masks){
```
Comment (wr-candidate) [inline src/validator/CreditCardVendor.java:22]: 1. Please reorder the methods (constructors come first). Modern IDE should take care of it as well.
2. Please use the same variable names for input arguments and fields.

File: src/validator/CreditCardValidatorProject.java [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+	private static CreditCardValidator validator = new CreditCardValidator();
+	private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+	
+	public static void main(String[] args) {
+        System.out.println("CREDIT CARD VALIDATOR\n");
+        
+        int vendorsLength = vendors.size();
+        int selected = 0;
+        
+        while (true){
```
Comment (wr-candidate) [inline src/validator/CreditCardValidatorProject.java:21]: Please move any application logic out of the `main` method into a separate class, so that `static` fields can be avoided.

File: src/validator/CreditCardValidator.java [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.ArrayList;
+
+public class CreditCardValidator implements CreditCardChecking {
+
+	@Override
+	public boolean checkValidity(CreditCardVendor ccv, String number) {
+		ArrayList<Integer> chars = ccv.getNumberOfChars();
+		ArrayList<String> masks = ccv.getIIN();
+		
+		boolean result1 = checkLengthCorrectness(number, chars);
```
Comment (wr-candidate) [inline src/validator/CreditCardValidator.java:12]: `final` and use more meaningful names rather than `resultN`.

File: src/validator/CreditCardValidator.java [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+		boolean result2 = checkIINMaskCorrectness (number, masks);
+		boolean result3 = checkLuhnalgorithmCorrectness(number);
+		
+		return result1 && result2 && result3;
+	}
+	
+	private boolean checkLengthCorrectness(String number, ArrayList<Integer> chars){
```
Comment (wr-candidate) [inline src/validator/CreditCardValidator.java:19]: Please extract this logic to separate classes so that unit tests can be created for them.

File: src/validator/CreditCardValidator.java [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	
+	private boolean checkLuhnalgorithmCorrectness(String number){
+		boolean result = false;
+		
+		String[] characters = number.split("");
+		int[] intCharsReverse = new int[characters.length];
+		int sum = 0;
+		
+		for(int i = 0;i<characters.length;i++){
```
Comment (wr-candidate) [inline src/validator/CreditCardValidator.java:44]: This steps can be extracted to private methods and named, so that it will be easier to reason about them later on.

File: src/validator/CreditCardValidator.java [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+		
+		for(int i = 0;i<intCharsReverse.length;i++){
+			intCharsReverse[i] = (i%2==1) ? 2*intCharsReverse[i] : intCharsReverse[i];
+		}
+		
+		for(int i = 0;i<intCharsReverse.length;i++){
+			sum += (intCharsReverse[i]<10) ? intCharsReverse[i] : (intCharsReverse[i]%10 + intCharsReverse[i]/10);
+		}
+		
+		if(sum%10==0)
```
Comment (wr-candidate) [inline src/validator/CreditCardValidator.java:56]: `10` is a magic number - please extract if to a class-level constant.

**Review by wr-candidate — COMMENTED** [review]
This PR needs work. In order to move forward the following should be discussed first and foremost:
- IDE used and formatting options (common to all engineers working on the project)
    - braces
    - naming convention
    - order of declarations
- using coding principles like SOLID or Clean Code
- NULLs handling
- Immutability (`final`s, immutable structures), and why it matters
- Dependency Injection
- Tests

Once the aforementioned topics are covered, more advanced topics should be agreed upon:
- build tool & dependency management
- HTTP Frameworks (the task was about creating REST API, but CLI was created instead)
- App configuration
- Logging
- Metrics
- Security (API Token / HTTPS 2way auth)
- Rate limits

Given the number of things to be tackled I would suggest setting up a meeting (or set of meetings).
